main/src/library/System.flix

```diff
@@ -30,8 +30,22 @@ namespace System {
         ///
         /// See also `Console.readLine` for reading from the console.
         ///
-        /// pub def readLines(): Iterator[String] & Impure = ???  // TODO
-
+        pub def readLines(): Iterator[String] & Impure =
+            import static get java.lang.System.in as getSystemIn;
+            import new java.io.InputStreamReader(##java.io.InputStream) as newInputStream;
+            import new java.io.BufferedReader(##java.io.Reader) as newBufferedReader;
+            import java.io.BufferedReader.readLine() as bfReadLine;
+
+            let br = newBufferedReader(newInputStream(getSystemIn()) as ##java.io.Reader); // up-cast to superclass
+            let nextLine = ref bfReadLine(br);
```

Comment:
So you can't do it like this because you ask for the first line before done or next has been called. You have to wait until done is called before doing the first read.

Reply:
So it is acceptable that done is reading ahead but readlines() isn't? I can do that.